File: Data Structures & Algorithms/target-sum/submission-3.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        dp = {} # key: (i, total), val = no. of ways till i, for total

        def dfs(i, total):
            if i == n:
                return 1 if total == target else 0
            if (i, total) in dp:
                return dp[(i, total)]
   
            dp[(i, total)] =(
                dfs(i + 1, total + nums[i]) + 
                dfs(i + 1, total - nums[i])
            )
            return dp[(i, total)]
        
        return dfs(0, 0)


        

File: Data Structures & Algorithms/majority-element/submission-3.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int cnt = 1;
        for(int i =1; i < n; i++){
            if(nums[i] == nums[i-1]){
                cnt++;
                if(cnt > (n/2)){
                    return nums[i];
                }
            }
            else{
                cnt = 1;
            }
        }
        return nums[n-1];
    }
};